### PR TITLE
BUG: iloc.__setitem__ with dict value and mixed dtypes

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/indexing/test_indexing.py [hunk 1]
```diff
@@ -741,13 +741,21 @@ def test_slice_with_zero_step_raises(self):
...
-        df = DataFrame({"x": [1, 2, 6], "y": [2, 2, 8], "z": [-5, 0, 5]}).set_index("z")
+        index = Index([-5, 0, 5], name="z")
+        df = DataFrame({"x": [1, 2, 6], "y": [2, 2, 8]}, index=index)
         expected = df.copy()
         rhs = {"x": 9, "y": 99}
         df.loc[5] = rhs
         expected.loc[5] = [9, 99]
         tm.assert_frame_equal(df, expected)
 
+        # GH#38335 same thing, mixed dtypes
```
Comment (jreback) [inline pandas/tests/indexing/test_indexing.py:752]: was this broken too? (or just iloc as indicated)?

is `[]` ok?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/indexing/test_indexing.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/tests/indexing/test_indexing.py:752]: loc is broken too.  setitem is OK i think.

File: pandas/core/indexing.py [hunk 2]
```diff
@@ -1649,8 +1649,10 @@ def _setitem_with_indexer_split_path(self, indexer, value, name: str):
         if isinstance(indexer[0], np.ndarray) and indexer[0].ndim > 2:
             raise ValueError(r"Cannot set values with ndim > 2")
 
-        if isinstance(value, ABCSeries) and name != "iloc":
-            value = self._align_series(indexer, value)
+        if (isinstance(value, ABCSeries) and name != "iloc") or isinstance(value, dict):
```
Comment (jreback) [inline pandas/core/indexing.py:1652]: maybe put an if clause before this
```
if isinstance(value, dict):
    value = Series(value)
```

and leave the original, as a bit more clear? (e.g. you know i like terser code but this compound and is treating very differnt cases); also we are boxing an already created Series with Series (nbd but not as clean)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexing.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/indexing.py:1652]: the trouble with that is then it wouldn't do the alignment if name == "iloc"

**Review by jreback — APPROVED** [review]

**Pull request merged**